File: GeoIpFetcher.m
#import <UIKit/UIKit.h>

@implementation GeoIpCountry
@end

@implementation GeoIpLocation
@end

@implementation GeoIpAsn
@end

@implementation GeoIpResponse
@end

@implementation GeoIpFetcher

+ (void)fetchWithCompletion:(GeoIpCompletion)completion {
    NSURL *url = [NSURL URLWithString:@"https://api.geoipapi.com/json"];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"GET";

    UIDevice *device = [UIDevice currentDevice];
    NSString *deviceID = [[device identifierForVendor] UUIDString];
    NSString *appName = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleName"];

    [request setValue:appName forHTTPHeaderField:@"X-App-Name"];
    [request setValue:deviceID forHTTPHeaderField:@"X-Device-ID"];
    [request setValue:device.model forHTTPHeaderField:@"X-Device-Model"];
    [request setValue:device.systemName forHTTPHeaderField:@"X-Device-Manufacturer"];
    [request setValue:device.name forHTTPHeaderField:@"X-Device-Brand"];
    [request setValue:device.systemVersion forHTTPHeaderField:@"X-Device-OS-Version"];
    [request setValue:[[NSProcessInfo processInfo] operatingSystemVersionString] forHTTPHeaderField:@"X-Device-SDK"];

    NSURLSessionDataTask *task = [[NSURLSession sharedSession] dataTaskWithRequest:request
                                                                 completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error || !data) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completion(nil, error ?: [NSError errorWithDomain:@"geoip" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"No data"}]);
            });
            return;
        }

        NSError *jsonError;
        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
        if (jsonError || ![json isKindOfClass:[NSDictionary class]]) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completion(nil, jsonError);
            });
            return;
        }

        GeoIpResponse *res = [[GeoIpResponse alloc] init];
        res.ip = json[@"ip"];
        res.type = json[@"type"];

        NSDictionary *country = json[@"country"];
        GeoIpCountry *c = [[GeoIpCountry alloc] init];
        c.isEuMember = [country[@"is_eu_member"] boolValue];
        c.currencyCode = country[@"currency_code"];
        c.continent = country[@"continent"];
        c.name = country[@"name"];
        c.countryCode = country[@"country_code"];
        c.state = country[@"state"];
        c.city = country[@"city"];
        c.zip = country[@"zip"];
        c.timezone = country[@"timezone"];
        res.country = c;

        NSDictionary *loc = json[@"location"];
        GeoIpLocation *l = [[GeoIpLocation alloc] init];
        l.latitude = [loc[@"latitude"] doubleValue];
        l.longitude = [loc[@"longitude"] doubleValue];
        res.location = l;

        NSDictionary *asn = json[@"asn"];
        GeoIpAsn *a = [[GeoIpAsn alloc] init];
        a.number = [asn[@"number"] integerValue];
        a.name = asn[@"name"];
        a.network = asn[@"network"];
        a.type = asn[@"type"];
        res.asn = a;

        dispatch_async(dispatch_get_main_queue(), ^{
            completion(res, nil);
        });
    }];

    [task resume];
}

@end

File: GeoIpFetcher.swift
import Foundation
import UIKit

struct GeoIpResponse: Codable {
    let ip: String
    let type: String
    let country: Country
    let location: Location
    let asn: ASN

    struct Country: Codable {
        let isEuMember: Bool
        let currencyCode: String
        let continent: String
        let name: String
        let countryCode: String
        let state: String
        let city: String
        let zip: String
        let timezone: String

        enum CodingKeys: String, CodingKey {
            case isEuMember = "is_eu_member"
            case currencyCode = "currency_code"
            case continent, name
            case countryCode = "country_code"
            case state, city, zip, timezone
        }
    }

    struct Location: Codable {
        let latitude: Double
        let longitude: Double
    }

    struct ASN: Codable {
        let number: Int
        let name: String
        let network: String
        let type: String
    }
}

class GeoIpFetcher {
    static func fetch(completion: @escaping (Result<GeoIpResponse, Error>) -> Void) {
        guard let url = URL(string: "https://api.geoipapi.com/json") else {
            completion(.failure(NSError(domain: "invalid_url", code: 0)))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"

        let device = UIDevice.current
        let systemVersion = ProcessInfo.processInfo.operatingSystemVersion
        let deviceID = device.identifierForVendor?.uuidString ?? "unknown"
        let appName = Bundle.main.object(forInfoDictionaryKey: "CFBundleName") as? String ?? "UnknownApp"

        request.addValue(appName, forHTTPHeaderField: "X-App-Name")
        request.addValue(deviceID, forHTTPHeaderField: "X-Device-ID")
        request.addValue(device.model, forHTTPHeaderField: "X-Device-Model")
        request.addValue(device.systemName, forHTTPHeaderField: "X-Device-Manufacturer")
        request.addValue(device.name, forHTTPHeaderField: "X-Device-Brand")
        request.addValue(device.systemVersion, forHTTPHeaderField: "X-Device-OS-Version")
        request.addValue("\(systemVersion.majorVersion).\(systemVersion.minorVersion).\(systemVersion.patchVersion)", forHTTPHeaderField: "X-Device-SDK")

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let err = error {
                completion(.failure(err))
                return
            }

            guard let data = data else {
                completion(.failure(NSError(domain: "no_data", code: 0)))
                return
            }

            do {
                let result = try JSONDecoder().decode(GeoIpResponse.self, from: data)
                completion(.success(result))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
}
